README.md
# ECE366Group11

- Place information revalent to delieverables and running code here.
- Our ISA design is called CTZ (Cam Teja Zach) it had the best ring to it.

## ISA Instruction Information

![alt text](https://github.com/lohe987/ECE366Group11/blob/master/ISA_Picture2.PNG)

## ECE366_mod17_7inst.asm

Believed to be working version for any P and Q but DIC is very high.

## Python program `assembler.py`

Execute: `python assembler.py`

- This python script will assemble instructions of the CTZ ISA design into binary machine code. 
- The program adds even parity bit encodeing as the first bit in each instructon.
- Defaultly uses CTZ_instructions.txt as input and CTZ_machine_code.txt as output when run. 


CHANGELOG.md
# ECE366Group11 Change Log

All notable changes are documented here.

Subheadings are the in format MM.DD.XX, MM is month, DD is day, XX is change for that day start index at 0.

Please add most recent updates first that way the list will be in reverse cornicallogical order.

## 10.08.00

- Changed instruction set to include set less than (slt)
- Jump now takes a register value to be determined if that is the line of code or the offset
- Add bitwise not command so no need to xor with 0xFFFF to invert.

## 10.07.00

- Finished `assembler.py` believed to be working version tested and complete.
- Try to start Par1_ISA but I think we really need set less than otherwise I dont see how we can do the loops
- I got stuck once I got to take the mod part

## 10.01.00

- Spent 1 hours starting assembler program works for about half of the instructions we picked
- Currently uses dictionaries to look up proper code and encode
- Praity bit is working testing no bugs found
- assemble_file uses default args of CTZ_instructions.txt and output of CTZ_machine_code.txt
- TO DO is 2 comps function change a number into 2's comp

## 09.30.00

- Spent 2 hours working on updating my 6^P mod Q code to use only 7 different instructions. DIC is very high.
- Added Change Log file
- Took MIPS code and put it in text file with line numbers as outlined in project documents.
- I think jump should only go back or to some imm for 0-15. Still need to rewrite Best match to be sure.


mips2.asm
.data
#T: .word 12
T: .word 0xABCDEF00
#T: .word -5
best_matching_score: .word -1 # best score = ? within [0, 32]
best_matching_count: .word -1 # how many patterns achieve the best score?
Pattern_Array: .word 0, 1, 2, 3, 4, -1, -2, -3, -4, -5, 0xEEEEEEEE, 0x44448888, 0x77777777, 0x33333333, 0xAAAAAAAA, 0xFFFF0000, 0xFFFF, 0xCCCCCCCC, 0x66666666, 0x99999999  
#Pattern_Array: .word 0, 1, 2, 3, 4, -1, -2, -3, -4, -5, 0xEEEEEEEE, 0x44448888, 0x77777777, 0x33333333, 0xAAAAAAAA, 0xFFFF0000, 0xFFFF, 0xCCCCCCCC, 0x66666666, 0x99999999
#Pattern_Array: .word 1, -2, 3, -4, 5, -6, 7, -8, 9, -10, -5, 5, -5, 5, -5, 1, -2, 3, -4, 5

.text
# Your code goes here
Main:
	addi $4,$4,20		#big counter $4
	addi $5,$5,0x200c	#address of the array
	addi $7,$7,1000
	j Number

Number:
	lw $3,T
	lw $6,($5)
	addi $11,$11,32
	add $12,$0,$0
	
Checker:	
	beq $11,$0,Max_checker
	andi $9,$3,0x1
	andi $10,$6,0x1
	beq $9,$10,shift
	addi $12,$12,1		# this hamming distance for the number
shift:	srl $3,$3,1
	srl $6,$6,1
	subi $11,$11,1
	j Checker
	
Max_checker:
	beq $7,$12,max_equal
	slt $13,$12,$7
	beq $13,0,shift_array
	add $7,$0,$12
	addi $8,$0,1
	j shift_array
	
max_equal:
	addi $8,$8,1
	j shift_array

shift_array:
	subi $4,$4,1
	beq  $4,0,exit
	addi $5,$5,4
	j Number

exit: 
	addi $19,$19,32
	sub  $19,$19,$7
	sw   $19, 0x2004($0)
	sw   $8,  0x2008($0)

Done: 
	j Done
	
	
		




simulator.py
import sys
import collections

# Class CPU will hold the information of the CPU
class CPU:
    PC = 0 # Program Counter
    DIC = 0 # Insturction Counter
    R = [0] * 4 # Register Values
    instructions = [] # instructions in array
    memory = [] # memory in array

def check_parity_bit(machine_line):
    # Count the number of zeros and ones
    one_zero_dict = collections.Counter(machine_line)
    # Make sure an even number of ones exist in the instructions
    if one_zero_dict["1"] % 2 == 0:
        return True
    return False

def load_program(cpu, instr_file_name, memory_file_name):
    instr_file = open(instr_file_name, "r")
    memory_file = open(memory_file_name, "r")

    for line in instr_file:
        line = line.strip()
        if len(line) < 1 or line.startswith("#"):
            continue
        cpu.instructions.append(line)

    for line in memory_file:
        line = line.strip()
        if len(line) < 1 or line.startswith("#"):
            continue
        number = int(line,2)
        if line.startswith("1"):
            number = (0xFFFF - int(line,2) + 1) * -1
        cpu.memory.append(number)

    for i in range(128-len(cpu.memory)):
        cpu.memory.append(0)

    instr_file.close()
    memory_file.close()

    return cpu

def run_program(cpu):
    finished = False
    while(not finished):
        instr = cpu.instructions[cpu.PC]
        if not check_parity_bit(instr):
            print("ERROR: Parity Bit Error")
            sys.exit()

        if instr[1:4] == "000":
            # Load instruction
            Rx = registers[instr[4:6]]
            Ry = registers[instr[6:8]]
            cpu.R[Rx] = cpu.memory[cpu.R[Ry]]
            cpu.PC = cpu.PC + 1
            cpu.DIC = cpu.DIC + 1
        elif instr[1:4] == "001":
            # Store instruction
            Rx = registers[instr[4:6]]
            Ry = registers[instr[6:8]]
            cpu.memory[cpu.R[Ry]] = cpu.R[Rx]
            cpu.PC = cpu.PC + 1
            cpu.DIC = cpu.DIC + 1
        elif instr[1:4] == "010":
            # Add instruction
            Rx = registers[instr[4:6]]
            Ry = registers[instr[6:8]]
            cpu.R[Rx] = cpu.R[Rx] + cpu.R[Ry]
            cpu.PC = cpu.PC + 1
            cpu.DIC = cpu.DIC + 1
        elif instr[1:4] == "101":
            # Init instruction
            Rx = registers[instr[4:6]]
            value = {"10" : -2,
                     "11" : -1,
                     "00" : 0,
                     "01" : 1}
            cpu.R[Rx] = value[instr[6:8]]
            cpu.PC = cpu.PC + 1
            cpu.DIC = cpu.DIC + 1
        elif instr[1:6] == "11010":
            # Jump
            Rx = registers[instr[6:8]]
            if cpu.R[Rx] == 0:
                finished = True
            cpu.PC = cpu.PC + cpu.R[Rx]
            cpu.DIC = cpu.DIC + 1
        elif instr[1:4] == "011":
            # Branch Equal
            Rx = registers[instr[4:6]]
            Ry = registers[instr[6:8]]
            if cpu.R[Rx] == cpu.R[Ry]:
                cpu.PC = cpu.PC + 1
            cpu.PC = cpu.PC + 1
            cpu.DIC = cpu.DIC + 1
        elif instr[1:4] == "100":
            # Set less than
            Rx = registers[instr[4:6]]
            Ry = registers[instr[6:8]]
            if cpu.R[Rx] < cpu.R[Ry]:
                cpu.R[0] = 1
            else
                cpu.R[0] = 0
            cpu.PC = cpu.PC + 1
            cpu.DIC = cpu.DIC + 1
        elif instr[1:4] == "111":
            # xor
            Rx = registers[instr[4:6]]
            Ry = registers[instr[6:8]]
            cpu.R[Rx] = cpu.R[Rx] ^ cpu.R[Ry] # May not work correctly because of way python does 2's comp
            cpu.PC = cpu.PC + 1
            cpu.DIC = cpu.DIC + 1
        else:
            print("Error Unknown command")
            sys.exit()
    return cpu

registers = {"00" : 0,
             "01" : 1,
             "10" : 2,
             "11" : 3}

if __name__ == "__main__":
    cpu1 = CPU()
    cpu1 = load_program(cpu1, "sample_program_machine.ctz", "sample_memory.txt")
    print(cpu1.memory[0:5])
    cpu1 = run_program(cpu1)
    print(cpu1.R)
    print(cpu1.DIC)
    print(cpu1.memory[0:5])


disassembler.py
import collections

def check_parity_bit(machine_line):
    # Count the number of zeros and ones
    one_zero_dict = collections.Counter(machine_line)
    # Make sure an even number of ones exist in the instructions
    if one_zero_dict["1"] % 2 == 0:
        return True
    return False

def determine_function(machine_line):
    if machine_line[1:4] == "110":
        return single_register(machine_line)
    elif machine_line[1:4] == "101":
        return imm_value(machine_line)
    else:
        return double_register(machine_line)

def single_register(machine_line):
    op_code = {"10" : "jump",
               "00" : "slr",
               "01" : "and",
               "11" : "not"}
    # Determine function code
    asm_line = op_code.get(machine_line[4:6], "UNKNOWN")
    # Add register values
    asm_line += " " + registers_bin_text.get(machine_line[6:8], "??")
    return asm_line

def imm_value(machine_line):
    value = {"11" : "-2",
             "10" : "-1",
             "00" : "0",
             "01" : "1"}
    # Add function word and register
    asm_line = "init" + " " + registers_bin_text.get(machine_line[4:6], "??") + ", "
    # Find imm value in table
    asm_line += value.get(machine_line[6:8], "??")
    return asm_line

def double_register(machine_line):
    func_code = {"000" : "load",
                 "001" : "store",
                 "010" : "add",
                 "011" : "beq",
                 "100" : "slt",
                 "111" : "xor"}
    # Add function code to begining of assembler line
    asm_line = func_code.get(machine_line[1:4], "UNKNOWN")
    # Update with register values
    asm_line += " " + registers_bin_text.get(machine_line[4:6], "??") + ", " + registers_bin_text.get(machine_line[6:8], "??")
    return asm_line

def disassemble_file(input_file_name="CTZ_machine_code.txt", output_file_name="CTZ_instructions_decoded.txt"):
    # Open files using default args or user provided
    input_file = open(input_file_name, "r")
    output_file = open(output_file_name, "w")

    # For each line in input file process the function
    for line in input_file:
        # Remove new line for line
        line = line.replace("\n", "")

        comment = ""
        asm_line = ""
        index = line.find("#")
        if index > -1:
            comment = line[index:]
            line = line[0:index]

        if len(line) > 0:
            if check_parity_bit(line):
                asm_line = determine_function(line)
            else:
                asm_line = "PARITY ERROR"

        output_file.write(asm_line + " " + comment + "\n")

    input_file.close()
    output_file.close()

registers_bin_text = {"00" : "R0",
                      "01" : "R1",
                      "10" : "R2",
                      "11" : "R3"}

if __name__ == "__main__":
    input_file_1 = "project2_group_11_p1_bin.txt"
    output_file_1 = "project2_group_11_p1_asm.txt"
    input_file_2 = "project2_group_11_p2_bin.txt"
    output_file_2 = "project2_group_11_p2_asm.txt"
    disassemble_file(input_file_2, output_file_2)


project2_group_11_dis.py
import collections

def check_parity_bit(machine_line):
    # Count the number of zeros and ones
    one_zero_dict = collections.Counter(machine_line)
    # Make sure an even number of ones exist in the instructions
    if one_zero_dict["1"] % 2 == 0:
        return True
    return False

def determine_function(machine_line):
    if machine_line[1:4] == "110":
        return single_register(machine_line)
    elif machine_line[1:4] == "101":
        return imm_value(machine_line)
    else:
        return double_register(machine_line)

def single_register(machine_line):
    op_code = {"10" : "jump",
               "00" : "slr",
               "01" : "and",
               "11" : "not"}
    # Determine function code
    asm_line = op_code.get(machine_line[4:6], "UNKNOWN")
    # Add register values
    asm_line += " " + registers_bin_text.get(machine_line[6:8], "??")
    return asm_line

def imm_value(machine_line):
    value = {"11" : "-2",
             "10" : "-1",
             "00" : "0",
             "01" : "1"}
    # Add function word and register
    asm_line = "init" + " " + registers_bin_text.get(machine_line[4:6], "??") + ", "
    # Find imm value in table
    asm_line += value.get(machine_line[6:8], "??")
    return asm_line

def double_register(machine_line):
    func_code = {"000" : "load",
                 "001" : "store",
                 "010" : "add",
                 "011" : "beq",
                 "100" : "slt",
                 "111" : "xor"}
    # Add function code to begining of assembler line
    asm_line = func_code.get(machine_line[1:4], "UNKNOWN")
    # Update with register values
    asm_line += " " + registers_bin_text.get(machine_line[4:6], "??") + ", " + registers_bin_text.get(machine_line[6:8], "??")
    return asm_line

def disassemble_file(input_file_name="CTZ_machine_code.txt", output_file_name="CTZ_instructions_decoded.txt"):
    # Open files using default args or user provided
    input_file = open(input_file_name, "r")
    output_file = open(output_file_name, "w")

    # For each line in input file process the function
    for line in input_file:
        # Remove new line for line
        line = line.replace("\n", "")

        comment = ""
        asm_line = ""
        index = line.find("#")
        if index > -1:
            comment = line[index:]
            line = line[0:index]

        if len(line) > 0:
            if check_parity_bit(line):
                asm_line = determine_function(line)
            else:
                asm_line = "PARITY ERROR"

        output_file.write(asm_line + " " + comment + "\n")

    input_file.close()
    output_file.close()

registers_bin_text = {"00" : "R0",
                      "01" : "R1",
                      "10" : "R2",
                      "11" : "R3"}

if __name__ == "__main__":
    input_file_1 = "Program1_machine_code.txt"
    output_file_1 = "Program1_disasm.txt"
    input_file_2 = "Part2_ISA_machine_code.txt"
    output_file_2 = "Program2_disasm.txt"
    disassemble_file(input_file_2, output_file_2)


assembler.py
import collections

def instr_load(line):
    # 000 RX RY
    if line[0] != "load":
        return "ERROR LINE CANNOT BE PARSED"
    else:
        return "000" + registers.get(line[1], "??") + registers.get(line[2], "??")

def instr_store(line):
    # 001 RX RY
    if line[0] != "store":
        return "ERROR LINE CANNOT BE PARSED"
    else:
        return "001" + registers.get(line[1], "??") + registers.get(line[2], "??")

def instr_add(line):
    # 010 RX RY
    if line[0] != "add":
        return "ERROR LINE CANNOT BE PARSED"
    else:
        return "010" + registers.get(line[1], "??") + registers.get(line[2], "??")

def instr_branch(line):
    # 011 RX RY
    if line[0] != "beq":
        return "ERROR LINE CANNOT BE PARSED"
    else:
        return "011" + registers.get(line[1], "??") + registers.get(line[2], "??")

def instr_jump(line):
    # 110 10 RR
    return "110" + "10" + registers.get(line[1], "??")

def instr_init(line):
    # 101 RR ii
    return "101" + registers.get(line[1], "??") + convert_imm_value(line[2])[-2:]

def instr_slr(line):
    # 110 00 RR
    return "110" + "00" + registers.get(line[1], "??")

def instr_and(line):
    # 110 01 RR
    return "110" + "01" + registers.get(line[1], "??")

def instr_slt(line):
    # 100 RX RY
    return "100" + registers.get(line[1], "??") + registers.get(line[2], "??")

def instr_xor(line):
    # 111 RR RR
    return "111" + registers.get(line[1], "??") + registers.get(line[2], "??")

def instr_not(line):
    # 110 11 RR
    return "110" + "11" + registers.get(line[1], "??")

def add_praity_bit(encoded_line):
    # Adds even parity bit to encoded_line
    # Count the number of zeros and ones in the encoded line
    one_zero_dict = collections.Counter(encoded_line)
    # If number of ones is odd at 1 as the parity bit
    if one_zero_dict["1"] % 2 == 1:
        return "1" + encoded_line
    else:
        return "0" + encoded_line

def convert_imm_value(number):
    if int(number) < 0:
        number = 0xFFFF - int(number[1:]) + 1
    return format(int(number), "016b")

def assemble_file(input_file_name="CTZ_instructions.txt", output_file_name="CTZ_machine_code.txt"):
    # Open files using default args or user provided
    input_file = open(input_file_name, "r")
    output_file = open(output_file_name, "w")

    # For each line in input file process the function
    for line in input_file:
        old_line = line
        comment = ""
        func = "?"
        if line.find("#") > -1:
            comment = line[line.find("#"):]
            comment = comment.strip("\n")

        # Remove comments from code
        line = line.strip()
        if line.startswith("#"):
            output_file.write(line + "\n")
        elif len(line) < 1:
            output_file.write("\n")
        else:
            line = line.replace(",", " ")
            line = line.replace("\n", "")
            line = line.replace("\t", " ")
            line = line.split(" ")
            line = list(filter(None, line))
            func = instructions.get(line[0], "?")
            # Check if function is known if not print error message
            if func == "?":
                output_file.write("UNKNOWN\n")
            else:
                # Encode function then add praity bit
                encoded = add_praity_bit(func(line))
                output_file.write(encoded + " " + comment + "\n")

    input_file.close()
    output_file.close()



instructions = {"load" : instr_load,
               "store" : instr_store,
               "add" : instr_add,
               "beq" : instr_branch,
               "jump" : instr_jump,
               "init" : instr_init,
               "and" : instr_and,
               "slt" : instr_slt,
               "slr" : instr_slr,
               "not" : instr_not,
               "xor" : instr_xor}

registers = {"R0" : "00",
             "R1" : "01",
             "R2" : "10",
             "R3" : "11",
             "r0" : "00",
             "r1" : "01",
             "r2" : "10",
             "r3" : "11"}


if __name__ == "__main__":
    input_file_name = "Program1.txt"
    output_file_name = "project2_group_11_p1_bin.txt"
    assemble_file(input_file_name, output_file_name)
